### PR TITLE
KEYCLOAK-16134 Allow webauthn idless login flow

### LINKED ISSUE #10832: Allow WebAuthn idless login flow
### Description

See https://github.com/keycloak/keycloak/pull/7860 and https://issues.redhat.com/browse/KEYCLOAK-16134

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by vanrar68**
Rework of a previously closed PR #7564 with changes requested by @mposolda
Prior to this PR WebAuthN authenticators (regular and passwordless) required a user to be set before asking for a webauthn device.
This PR removes the need for a user to be previously set when a webauthn credential containing a user id is submitted.

### CODE REVIEW

**Comment by vanrar68** [comment]
Added some modifications of the authenticator setup code in DefaultAuthenticationFlow.java. 
Previously authenticator setup was only possible for authenticators that required users to be set (authenticator.requiresUser() == true). Now we also need to allow authenticator setup for authenticators that don't always require the user to be set.

**Comment by mposolda** [comment]
@vanrar68 Sorry for late response. To me, the changes looks good.

TBH I would prefer to let them in once we have WebAuthn automated tests in place to make sure that our currently supported WebAuthn use-cases (with WebAuthn as 2nd-factor or passwordless) are not somehow broken by this change.

@tnorimat Does this change looks good to you or do you see any further issues with it?

**Comment by mposolda** [comment]
@vanrar68 It will be also good to have some documentation for this. If you have an example how to integrate Keycloak with WebAuthn idless login, it will be nice to be added to the WebAuthn section of Keycloak documentation. Ideally details like:
- How to configure Keycloak authentication flow and WebAuthn policies together with some info etc (Ideally with the usage of WebAuthn passwordless policy)
- Which WebAuthn key you used for testing for idless (version, setup of the key if applicable etc).

**Comment by tnorimat** [comment]
Hello @vanrar68 , @mposolda 
I will review this PR today.

File: services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java
Comment (tnorimat) [inline services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java:435]: It seems that L435 and L436 changes the original codes' logic for condition of throwing `AuthenticationFlowException` in L442.

To keep the same condition for throwing it in the original codes, how about the following?

```
    if (authUser != null && !authenticator.configuredFor(processor.getSession(), processor.getRealm(), authUser)) {
        if (factory.isUserSetupAllowed() && model.isRequired() && authenticator.areRequiredActionsEnabled(processor.getSession(), processor.getRealm())) {
```

If you intentionally change this condition, could you tell me this reason?

File: services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java
Comment (tnorimat) [inline services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java:440]: `processor.getAuthenticationSession().getAuthenticatedUser()` can be replaced with `authUser` .

**Review by tnorimat — COMMENTED** [review]
@vanrar68 I've added some review comments. Could you check them?

**Comment by tnorimat** [comment]
@vanrar68 As @mposolda has mentioned, the arquillian integration test for this ID-less WebAuthn login flow is needed.

File: services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java
Comment (vanrar68) [reply services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java:435]: The purpose of the modification is to handle credential setup for authenticators that don't require a user to be set. Your proposition looks good to me, I'll update the code accordingly in order to stay the closest possible to the existing code (though I think we should get rid of the meaningless comment line)

File: services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java
Comment (vanrar68) [reply services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java:440]: Yep, will do

**Comment by vanrar68** [comment]
> @vanrar68 As @mposolda has mentioned, the arquillian integration test for this ID-less WebAuthn login flow is needed.

Yes I agree, but as mentioned by @mposolda there seem to be some major work going on right now on webauthn testing framework, see #8534. I'd rather wait until this work is merged and implement the loginless test on top of that even if this means waiting some more to have loginless support in Keycloak. What do you think ?

**Comment by vanrar68** [comment]
> @vanrar68 It will be also good to have some documentation for this. If you have an example how to integrate Keycloak with WebAuthn idless login, it will be nice to be added to the WebAuthn section of Keycloak documentation. Ideally details like:
> 
> * How to configure Keycloak authentication flow and WebAuthn policies together with some info etc (Ideally with the usage of WebAuthn passwordless policy)
> * Which WebAuthn key you used for testing for idless (version, setup of the key if applicable etc).

Yes, I'll create a PR for keycloak-documentation regarding the configuration of webauthn idless auth

File: services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java
Comment (tnorimat) [reply services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java]: @vanrar68 Thank you. I've checked your update.

File: services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java
Comment (tnorimat) [reply services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java]: @vanrar68 Thank you. I've checked your update.

**Comment by tnorimat** [comment]
@vanrar68 

https://github.com/keycloak/keycloak/pull/7860#issuecomment-971473482

Yes, I agree with you.

**Comment by vanrar68** [comment]
> @vanrar68 It will be also good to have some documentation for this. If you have an example how to integrate Keycloak with WebAuthn idless login, it will be nice to be added to the WebAuthn section of Keycloak documentation. Ideally details like:
> 
> * How to configure Keycloak authentication flow and WebAuthn policies together with some info etc (Ideally with the usage of WebAuthn passwordless policy)
> * Which WebAuthn key you used for testing for idless (version, setup of the key if applicable etc).

@mposolda Does this need a JIRA/Github Issue to be opened on the keycloak-documentation project or is a PR enough ?

**Comment by mposolda** [comment]
> > @vanrar68 As @mposolda has mentioned, the arquillian integration test for this ID-less WebAuthn login flow is needed.
> 
> Yes I agree, but as mentioned by @mposolda there seem to be some major work going on right now on webauthn testing framework, see #8534. I'd rather wait until this work is merged and implement the loginless test on top of that even if this means waiting some more to have loginless support in Keycloak. What do you think ?

+1 from me as well

**Comment by mposolda** [comment]
> > @vanrar68 It will be also good to have some documentation for this. If you have an example how to integrate Keycloak with WebAuthn idless login, it will be nice to be added to the WebAuthn section of Keycloak documentation. Ideally details like:
> > 
> > * How to configure Keycloak authentication flow and WebAuthn policies together with some info etc (Ideally with the usage of WebAuthn passwordless policy)
> > * Which WebAuthn key you used for testing for idless (version, setup of the key if applicable etc).
> 
> @mposolda Does this need a JIRA/Github Issue to be opened on the keycloak-documentation project or is a PR enough ?

PR is enough. You can use the same JIRA in this PR to keycloak-documentation as you used in this PR.

**Comment by mposolda** [comment]
@vanrar68 @tnorimat Thanks for the review Takashi and Thanks Joaquim for incorporating Takashi's feedback!

**Comment by vanrar68** [comment]
@mposolda, @tnorimat 
There's something else that would be interesting to have in Keycloak regarding loginless auth: the possibility for the user to select the auth method he wants to use **before** having to enter a username. 
Currently, my understanding is that if the auth flow is configured with alternative "executions" **AND** the user has already  submitted his username then Keycloak allows the user to select which of the alternative methods he wants to use (with the "Try another way" link) because the code in DefaultAuthenticationFlow.java builds a list of the available credentials for this user and matches these credentials with the authentication flow configuration.
With loginless auth it would be interesting to give the user the possibility to change the method of authentication ("Try another way") without having to enter a username. This way a user could choose between a Username/Password form or a WebAuthn Loginless auth. What do you think ?

**Comment by vanrar68** [comment]
> @vanrar68 @tnorimat Thanks for the review Takashi and Thanks Joaquim for incorporating Takashi's feedback!
> 
> However I see there are failing tests and looks it may be regression of this PR. This may need to be investigated. I guess that is the consequence of the changes in the DefaultAuthenticationFlow. If it is possible to address this PR without changes in DefaultAuthenticationFlow, it will be great (Not 100% sure it is possible).

I'll have a look

**Comment by vanrar68** [comment]
> @vanrar68 @tnorimat Thanks for the review Takashi and Thanks Joaquim for incorporating Takashi's feedback!
> 
> However I see there are failing tests and looks it may be regression of this PR. This may need to be investigated. I guess that is the consequence of the changes in the DefaultAuthenticationFlow. If it is possible to address this PR without changes in DefaultAuthenticationFlow, it will be great (Not 100% sure it is possible).

LOL, actually there is a way to have webauthn loginless without touching DefaultAuthenticationFlow: [PR7564](https://github.com/keycloak/keycloak/pull/7564)
The reason behind having to touch DefaultAuthenticationFlow is because we are diverting the "user requiring" WebAuthnPasswordless credential type to support Loginless auth (by making it not require a user anymore). 

I'll have a look at the failing tests anyway

**Comment by mposolda** [comment]
> @mposolda, @tnorimat
> There's something else that would be interesting to have in Keycloak regarding loginless auth: the possibility for the user to select the auth method he wants to use **before** having to enter a username.
> Currently, my understanding is that if the auth flow is configured with alternative "executions" **AND** the user has already submitted his username then Keycloak allows the user to select which of the alternative methods he wants to use (with the "Try another way" link) because the code in DefaultAuthenticationFlow.java builds a list of the available credentials for this user and matches these credentials with the authentication flow configuration.
> With loginless auth it would be interesting to give the user the possibility to change the method of authentication ("Try another way") without having to enter a username. This way a user could choose between a Username/Password form or a WebAuthn Loginless auth. What do you think ?

@vanrar68  FYI. There are already ways to authenticate before entering a username. For example Kerberos or X509 authentication login. However these are a bit different than WebAuthn Passwordless due the fact that there is no form.

Regarding the "WebAuthn Idless" login, it will be ideal, that if it is configured as ALTERNATIVE in the level 1 (same level of the authentication flow as "Cookie", "Kerberos" and "Form"), the user will see the screen with WebAuthn Passwordless as first screen. Then in case, he does NOT want to login with WebAuthn, but rather with username/password, he can see "Try another way" and click on it to login with Username/Password. However I am not 100% sure this is currently available without some major updates to DefaultAuthenticationFlow.

**Comment by mposolda** [comment]
@vanrar68 One more point: I think it will be more proper solution if method `Authenticator.requiresUser()` from Authenticator SPI has access to AuthenticationFlowContext. I think the fact that `requiresUser` does not have access to AuthenticationFlowContext is not good and it is quite a limitation.

AuthenticationFlowContext has access to the realm and hence WebAuthnPasswordless authenticator will be able to read the policy configuration and check if it is configured with "requireUser=true" and based on that, return true or false from the method `Authenticator.requiresUser(context)` . 

This will allow that method `Authenticator` itnerface can add new default method and change existing method to something like this:
```
// TODO javadoc
default boolean requiresUser(AuthenticationFlowContext context) {
    // Fallback to the old version of the method, so we don't need to change signatures of all the methods
    return requiresUser();
}

// TODO javadoc
@deprecated("Please use requiresUser(context) instead")
default boolean requiresUser() {
    return false;
}
```
Then the only change needed in the DefaultAuthenticationFlow would be to change the line with calling `requireUser` to the `requireUser(context)` . WDYT?

**Comment by vanrar68** [comment]
> @vanrar68 One more point: I think it will be more proper solution if method `Authenticator.requiresUser()` from Authenticator SPI has access to AuthenticationFlowContext. I think the fact that `requiresUser` does not have access to AuthenticationFlowContext is not good and it is quite a limitation.
> 
> AuthenticationFlowContext has access to the realm and hence WebAuthnPasswordless authenticator will be able to read the policy configuration and check if it is configured with "requireUser=true" and based on that, return true or false from the method `Authenticator.requiresUser(context)` .
> 
> This will allow that method `Authenticator` itnerface can add new default method and change existing method to something like this:
> 
> ```
> // TODO javadoc
> default boolean requiresUser(AuthenticationFlowContext context) {
>     // Fallback to the old version of the method, so we don't need to change signatures of all the methods
>     return requiresUser();
> }
> 
> // TODO javadoc
> @deprecated("Please use requiresUser(context) instead")
> default boolean requiresUser() {
>     return false;
> }
> ```
> 
> Then the only change needed in the DefaultAuthenticationFlow would be to change the line with calling `requireUser` to the `requireUser(context)` . WDYT?

Not sure I get this right. With only the "authenticator.requiresUser(context)" change made to DefaultAuthenticationFlow:
- if WebauthnPasswordless.requiresUser(context) returns true: LoginLess won't be possible
- if WebauthnPasswordless.requiresUser(context) returns false: Some login use cases won't trigger WebauthnPasswordless setup

**Comment by vanrar68** [comment]
> > @vanrar68 @tnorimat Thanks for the review Takashi and Thanks Joaquim for incorporating Takashi's feedback!
> > However I see there are failing tests and looks it may be regression of this PR. This may need to be investigated. I guess that is the consequence of the changes in the DefaultAuthenticationFlow. If it is possible to address this PR without changes in DefaultAuthenticationFlow, it will be great (Not 100% sure it is possible).
> 
> LOL, actually there is a way to have webauthn loginless without touching DefaultAuthenticationFlow: [PR7564](https://github.com/keycloak/keycloak/pull/7564) The reason behind having to touch DefaultAuthenticationFlow is because we are diverting the "user requiring" WebAuthnPasswordless credential type to support Loginless auth (by making it not require a user anymore).
> 
> I'll have a look at the failing tests anyway

@mposolda well, turns out I was wrong. I don't see WebAuthN LoginLess working without touching DefaultAuthenticationFlow (even with your code suggestion, see quoted reply). Reason is quite simple:
- WebAuthnPasswordLess credential needs to be flagged as not requiring user (obiously, otherwise LoginLess won't be possible)
- DefaultAuthenticationFlow currently prevents authenticator setup for some special use cases

BUT: the DefaultAuthenticationFlow code change included in this PR only affects those special use cases that require credential setup when no required action was previously set (eg. Username/Password form REQUIRED + OTP REQUIRED when no OTP is configured for the user).
My previous commit had the wrong checks for handling those use cases that's why 2 "broker link account" tests were failing. I fixed the code and included some detailed comments that should make things more clear. All tests are now OK

**Comment by vanrar68** [comment]
> > @mposolda, @tnorimat
> > There's something else that would be interesting to have in Keycloak regarding loginless auth: the possibility for the user to select the auth method he wants to use **before** having to enter a username.
> > Currently, my understanding is that if the auth flow is configured with alternative "executions" **AND** the user has already submitted his username then Keycloak allows the user to select which of the alternative methods he wants to use (with the "Try another way" link) because the code in DefaultAuthenticationFlow.java builds a list of the available credentials for this user and matches these credentials with the authentication flow configuration.
> > With loginless auth it would be interesting to give the user the possibility to change the method of authentication ("Try another way") without having to enter a username. This way a user could choose between a Username/Password form or a WebAuthn Loginless auth. What do you think ?
> 
> @vanrar68 FYI. There are already ways to authenticate before entering a username. For example Kerberos or X509 authentication login. However these are a bit different than WebAuthn Passwordless due the fact that there is no form.
> 
> Regarding the "WebAuthn Idless" login, it will be ideal, that if it is configured as ALTERNATIVE in the level 1 (same level of the authentication flow as "Cookie", "Kerberos" and "Form"), the user will see the screen with WebAuthn Passwordless as first screen. Then in case, he does NOT want to login with WebAuthn, but rather with username/password, he can see "Try another way" and click on it to login with Username/Password. However I am not 100% sure this is currently available without some major updates to DefaultAuthenticationFlow.

I checked how things are configured for Kerberos and X509. These authenticators differ from WebAuthn because they are not implementing CredentialValidator (ie. they don't rely on credentials stored in Keycloak DB). Authenticators that implement CredentialValidator currently can't be added to the AuthenticationSelectionOptions list if no user is associated to the current session (see AuthenticationSelectionResolver.createAuthenticationSelectionList()). Thus WebAuthn is currently not added to the list and the "Try another way" link is not presented to the user. 
I managed to make it work with a small change in AuthenticationSelectionResolver.createAuthenticationSelectionList() but need to run some more tests before submitting anything.

**Comment by vanrar68** [comment]
@mabartos I'm currently trying to implement webauthn idless tests by adapting your latest WebAuthn/WebAuthnPasswordless tests.
I see you have implemented ResidentKeyRegisterTest.residentKeyRequiredCorrect() but it's flagged as "Not working". Does this mean that residentkey webauthn test is not possible with keycloak current main branch ?

**Comment by mabartos** [comment]
@vanrar68 Those tests are not polished yet and deeper investigation will be done as some follow-up task. I wasn't able to execute the test for the Required requirements for the Resident Key. Either it's an issue of the test case itself or our implementation of the WebAuthn is wrong. I need to check it lately. However, you can try it with your custom test cases for this task and you'll see.

**Comment by vanrar68** [comment]
@mabartos both idless and passwordless auth scenarios are working fine with resident key set as required so I doubt the problem is coming from the WebAuthn implementation. The only thing not working as expected are my tests. I'll investigate deeper. Anyway thanks for your feedback

**Comment by vanrar68** [comment]
> @mabartos both idless and passwordless auth scenarios are working fine with resident key set as required so I doubt the problem is coming from the WebAuthn implementation. The only thing not working as expected are my tests. I'll investigate deeper. Anyway thanks for your feedback

@mabartos nevermind, I managed to make it work

**Comment by vanrar68** [comment]
@mposolda I've updated the logic in order to have a minimum impact on the existing code and added the tests.
Test runs seem ok. Let me know

**Comment by mabartos** [comment]
@vanrar68 LGTM! I've executed the pipeline tests and they passed [1]. 

I've been just thinking a little about the approach with the ID-less stuff. What about to move the ID-less option to the init login page same like the Identity providers? Instead of displaying the 'Try another way' there. I mean, it'd separate the authenticators by the fact whether the user is identified or not. 

Something like that:
idLess

@mposolda @vanrar68 WDYT? It's just and idea/consideration and I don't know what should be done in order to achieve that.

[1] https://keycloak-jenkins.com/job/universal-test-pipeline-server/1098/testReport/org.keycloak.testsuite.webauthn/WebAuthnIdlessTest/

**Comment by vanrar68** [comment]
> @vanrar68 LGTM! I've executed the pipeline tests and they passed [1].

Thanks, that's good news.
 
> I've been just thinking a little about the approach with the ID-less stuff. What about to move the ID-less option to the init login page same like the Identity providers? Instead of displaying the 'Try another way' there. I mean, it'd separate the authenticators by the fact whether the user is identified or not. Moreover, use the 'Try another way' only in the case the user is identified.

I'm definitely not the right person for UI advice/comment but I wonder what it would look like with multiple identity providers configured

**Comment by mabartos** [comment]
> I'm definitely not the right person for UI advice/comment but I wonder what it would look like with multiple identity providers configured

@vanrar68 Multiple IdPs. IMHO, I'd say this would be better approach, but I'm not sure. Maybe @mposolda will share his feedback for this.

img

**Comment by vanrar68** [comment]
> @vanrar68 Multiple IdPs. IMHO, I'd say this would be better approach, but I'm not sure. Maybe @mposolda will share his feedback for this.

IDPs are displayed on the login page whatever the auth flow is IIRC. Regarding the use of idless webauthn the plan was to only include it in the "Try another way" if the auth flow includes a webauthn-passwordless execution . Anyway, let's wait for more feedback

**Comment by vanrar68** [comment]
@mposolda Can you please have a look at the changes made on this PR and let me know if something's wrong ? Thanks

**Comment by mposolda** [comment]
@vanrar68 Yup, I had some other things, but I hope to take a deeper look this or next week. Sorry for the delay and Thanks for your patience and for your work on this

**Comment by mposolda** [comment]
Documentation PR: https://github.com/keycloak/keycloak-documentation/pull/1306

**Comment by mabartos** [comment]
@vanrar68 Thanks for the contribution! Before I review this PR, I'd like to discuss possible UI changes and improve usability of that. I created a discussion(#9821), so it'd be great if you'd share your thoughts or ideas.

**Comment by vanrar68** [comment]
Includes an attempt at moving the "Try another way" link on the login page (before Identity Providers buttons).
@mabartos can you please have a look ?

**Comment by vanrar68** [comment]
The failed tests are linked to the UI change not the idless feature itself

**Comment by mabartos** [comment]
@vanrar68 Thanks for the update. I'll check it.

**Comment by vanrar68** [comment]
This is my best (and last) shot at the UI:

try-another-way-1

try-another-way-2

try-another-way-3

**Comment by vanrar68** [comment]
@mabartos did you get a chance to review to UI part of this PR ?

File: themes/src/main/resources/theme/base/login/template.ftl
Comment (mabartos) [inline themes/src/main/resources/theme/base/login/template.ftl:128]: Could you please completely remove the `showAnotherWayIfPresent` property in the _template.ftl_?

File: services/src/main/java/org/keycloak/authentication/AuthenticationSelectionResolver.java
Comment (mabartos) [inline services/src/main/java/org/keycloak/authentication/AuthenticationSelectionResolver.java:98]: I was wondering if it's a proper solution to obtain `sessionFactory` in each iteration as the session (AFAIK) should be the same for the whole method. However, as one of the main purposes of the `AuthenticationProcessor` class is to manage the state of the authentication process and there are no other calls to dynamically changing objects, I don't consider this as a blocker for this PR. Just wanted to take it into consideration.

**Review by mabartos — COMMENTED** [review]
@vanrar68 Great work! In general, LGTM. 

Just one nitpick with the parameter in the `template.ftl` file, which could be solved (see below). The design changes look good. The test itself looks good as well, but it needs some polishing, regarding new changes in the WebAuthn testsuite, in order to make it stable for a long time, but IMO, it might be solved in some follow-up task, where I can solve it. @mposolda WDYT?

I executed the pipeline tests for WebAuthn and they passed. 
See https://keycloak-jenkins.com/job/universal-test-pipeline-server/1290/

(We need to focus only on the results for the specific test and browser Chrome... I verify the other failures are expected for this PR.)

@vanrar68 Thanks!

File: themes/src/main/resources/theme/base/login/template.ftl
Comment (vanrar68) [reply themes/src/main/resources/theme/base/login/template.ftl:128]: Done!

**Review by mabartos — DISMISSED** [review]
@vanrar68 Thanks! LGTM

@tnorimat It'd be wonderful if you'd have a chance to review it as well? 

And we need to wait for @mposolda approval.

**Comment by tnorimat** [comment]
@vanrar68 @mabartos Yes, I would like to review this PR.

**Review by tnorimat — COMMENTED** [review]
@vanrar68 I've added one tiny comment. Could you check it?

File: testsuite/integration-arquillian/tests/other/webauthn/src/test/java/org/keycloak/testsuite/webauthn/WebAuthnIdlessTest.java
Comment (vanrar68) [reply testsuite/integration-arquillian/tests/other/webauthn/src/test/java/org/keycloak/testsuite/webauthn/WebAuthnIdlessTest.java]: Done!

**Review by mabartos — APPROVED** [review]

**Comment by vanrar68** [comment]
@mposolda did you get a chance to review this PR ?

**Review by mposolda — APPROVED** [review]
@vanrar68 Did the review and testing and it looks good to me. Thanks for all your work on this!

**Comment by mposolda** [comment]
@mabartos @tnorimat Thanks for your review and feedback!

**Pull request merged**
